wwwapp/auth.py

```diff
@@ -41,154 +32,74 @@ def login_view(request):
-        if 'given_name' in user_info:
-            user_info['first_name'] = user_info['given_name']
-        elif 'name' in user_info and 'givenName' in user_info['name']:
-            user_info['first_name'] = user_info['name']['givenName']
-    if 'last_name' not in user_info:
-        if 'family_name' in user_info:
-            user_info['last_name'] = user_info['family_name']
-        elif 'name' in user_info and 'familyName' in user_info['name']:
-            user_info['last_name'] = user_info['name']['familyName']
-    if 'email' not in user_info:
-        if 'emails' in user_info and user_info['emails'] and 'value' in user_info['emails'][0]:
-            user_info['email'] = user_info['emails'][0]['value']
-
-
-# Extend django-all-access' OAuthRedirect to make G+ and emails work.
-class ScopedOAuthRedirect(OAuthRedirect):
-    def get_callback_url(self, provider):
-        return reverse('scopedallaccess-callback', kwargs={'provider': provider.name})
-
-    def get_additional_parameters(self, provider):
-        if provider.name == 'facebook':
-            return {'scope': 'public_profile email'}
-        if provider.name == 'google':
-            return {'scope': 'openid profile email'}
-        return super(ScopedOAuthRedirect, self).get_additional_parameters(provider)
-
-
-# Extend django-all-access' OAuthCallback to detect and merge duplicates.
-class ScopedOAuthCallback(OAuthCallback):
-    def handle_new_user(self, provider, access, info):
-        context = get_context(self.request)
-        # Check for users with the same emails or names.
-        standarize_user_info(info)
-        match_users = []
+@partial_step(save_to_session=False)
+def merge_accounts(strategy, details, request, response, current_partial, user=None, is_new=False, *args, **kwargs):
+    if is_new:
+        if user:
+            # This happens in 2 cases - when an already logged in user logs in with a different provider
+            # or when someone confirmed the ownership of an account
+            return {
+                'user': user,
+                'is_new': False,
+            }
+
+        new_backend = current_partial.backend
+        context = get_context(request)
         context['allow_account_creation'] = True
-        context['new_provider'] = provider
+        context['new_provider'] = new_backend
+        context['partial_token'] = current_partial.token
+        match_users = []
 
-        if 'email' in info and info['email']:
-            context['email'] = info['email']
-            match_users = list(User.objects.filter(email=info['email']))
-            if match_users:
+        email = details.get('email')
+        if email:
+            context['email'] = email
+            match_users = list(strategy.storage.user.get_users_by_email(email))
+            if len(match_users) == 1 and response.get('email_verified'):
+                # Don't perform additional validation when the mail was validated by google
```

Comment:
I thought we already discussed this? NO CHANGING OF CURRENT MERGING BEHAVIOR.

Reply:
And you are comparing the email with user-configurable value :joy:

Reply:
Still, please provide one example when things will go horribly wrong. If the user configured email X and someone logged in with a google verified email address X this means that they must have been the same person.
If a normal user changed the email to X then they still won't be able to hijack the account unless they have access to the email X.

CC: @cytadela8

Reply:
Sure, here you go

Person X does not have an account yet. Person Y wants to troll him, and knows that he is going to register for WWW this year. So person Y registers using Facebook, and changes his email to person X's Google email. When X registers with Google, his account is silently merged into the other one because the email matches and is verified, but Y still has a valid method of logging into that account.

Even if merging is not silent, the user may think "oh, I must have logged in with Facebook before and I just don't remember" and confirm the merge anyway.

In the old system, this would be impossible, as they would have to log in to the original account.